### PR TITLE
chore(typescript): convert library to typescript

### PR DESCRIPTION
**Pull request by mariusandra**
Needs https://github.com/PostHog/posthog-js/pull/428

## Changes

- Aiming to be a 100% compatible refactor with no changes to behaviour, and minimal changes to code

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers
- [ ] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

### CODE REVIEW

**Comment by mariusandra** [comment]
After more than a thousand manually squashed errors, posthog-js is now in TypeScript! All tests pass, and the library builds. The only thing to still check are the `.d.ts` files. So we're actually not there-there, but close.

I'll go through the PR and manually annotate less obvious changes tomorrow/today. I tried to leave all bugs intact, but sometimes it was easier to fix a thing or delete a feature we're not using than to type it.

There are several `any`-s around as well, especially around the complicated "same code works everywhere" browser initialization path.

File: src/posthog-people.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                    props[k] = v
+                }
+            })
+        } else {
+            props[prop] = to
+        }
+
+        data[actionType] = props
+        return data
+    }
```
Comment (mariusandra) [inline src/posthog-people.ts:173]: Moved here from api-actions.ts

File: src/posthog-persistence.ts [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    SESSION_RECORDING_ENABLED_SERVER_SIDE,
+    SESSION_ID,
+    ENABLED_FEATURE_FLAGS,
+]
+
+/**
+ * PostHog Persistence Object
+ * @constructor
+ */
+export class PostHogPersistence {
```
Comment (mariusandra) [inline src/posthog-persistence.ts:46]: Also converted to a class.

File: src/send-request.ts [hunk 3]
```diff
@@ -117,11 +123,7 @@ export const xhr = ({
                 }
 
                 if (callback) {
-                    if (options.verbose) {
-                        callback({ status: 0, error: error })
-                    } else {
-                        callback(0)
-                    }
+                    callback({ status: 0 })
```
Comment (mariusandra) [inline src/send-request.ts:126]: This fixes feature flags sending `[]` when decide errors.

https://github.com/PostHog/posthog-js/pull/420/files#diff-ddd24b8ce37c6ae2630e5b9f5a3a492bc0f16beabfa27858c3e19ed43a97d67c

🙈

File: src/types.ts [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+export type FeatureFlagsCallback = (flags: string[], variants: Record<string, string | boolean>) => void
+
+// TODO: delete custom_properties after changeless typescript refactor
+export interface AutoCaptureCustomProperty {
+    name: string
+    css_selector: string
+    event_selectors: string[]
+}
```
Comment (mariusandra) [inline src/types.ts:231]: Seems like a leftover from earlier times.

File: src/utils.ts [hunk 5]
```diff
@@ -87,71 +80,67 @@ var logger = {
...
-        ctor.prototype = func.prototype
-        var self = new ctor()
-        ctor.prototype = null
-        var result = func.apply(self, args.concat(slice.call(arguments)))
-        if (Object(result) === result) {
-            return result
-        }
-        return self
-    }
-    return bound
```
Comment (mariusandra) [inline src/utils.ts:113]: Removed since we don't support IE8, only IE11+, which supports this natively.

File: src/utils.ts [hunk 6]
```diff
@@ -87,71 +80,67 @@ var logger = {
...
+            }
+        }
+    }
+}
+
+export function _eachArray<E = any>(
+    obj: E[] | null | undefined,
+    iterator: (value: E, key: number) => void | Breaker,
+    thisArg?: any
+): void {
```
Comment (mariusandra) [inline src/utils.ts:127]: Making a separate each for arrays made typing much easier.

**Review by mariusandra — COMMENTED** [review]
Still some integration tests to fix, but y'all can already take a look :).

File: .github/workflows/integration.yml [hunk 7]
```diff
@@ -11,9 +11,9 @@ jobs:
...
       - uses: actions/setup-node@v1
         with:
-          node-version: 14
+          node-version: 16
 
       - run: yarn install && yarn build
 
       - name: Cypress run
-        uses: cypress-io/github-action@v2
+        uses: cypress-io/github-action@v4
```
Comment (mariusandra) [inline .github/workflows/integration.yml:19]: Part of upgrading Cypress to v10

File: src/gdpr-utils.ts [hunk 8]
```diff
@@ -237,64 +197,74 @@ function _optInOut(optValue, token, options) {
...
         // only capture event if opting in (optValue=true)
-        options.capture(options.captureEventName || '$opt_in', options.captureProperties, {
-            send_immediately: true,
+        options.capture(options.captureEventName || '$opt_in', options.captureProperties || {}, {
+            send_instantly: true,
         })
     }
 }
 
+export function userOptedOut(posthog: PostHog, silenceErrors: boolean | undefined) {
```
Comment (mariusandra) [inline src/gdpr-utils.ts:213]: This is extracted from `addOptOutCheck`, making it possible to check if we need to not capture without wrapping a function with magic.

File: src/posthog-core.ts [hunk 9]
```diff
@@ -0,0 +1,1711 @@
...
+    ): CaptureResult | void {
+        // While developing, a developer might purposefully _not_ call init(),
+        // in this case, we would like capture to be a noop.
+        if (!this.__loaded) {
+            return
+        }
+
+        if (userOptedOut(this, false)) {
+            return
+        }
```
Comment (mariusandra) [inline src/posthog-core.ts:657]: This is new. We used to wrap this function with `addOptOutCheck`

**Review by mariusandra — COMMENTED** [review]
All tests pass! Other than `TestCafe`, which seems to be flakey in general. Sometimes one of the three tests fails. Sometimes it's Safari. Sometimes Chrome.

**Comment by mariusandra** [comment]
Testing locally, the new library seems to work well with posthog itself. So, ready for review 💪

**Comment by benjackwhite** [comment]
Testcafe is unhappy but it actually does work on each browser. Tested it locally with browserstack and it was fine

**Pull request merged**